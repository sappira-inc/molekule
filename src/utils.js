import { get, kebabCase } from 'lodash';
import styled from 'styled-components';

export const themeGet = (theme, lookup, fallback) => get(theme, lookup, fallback);

export const getComponentVariant = (theme, componentName, variant) => {
  const config = themeGet(
    theme,
    `components.${componentName}.variants.${variant}`,
    theme.variants[componentName][variant]
  );
  if (!config) throw new Error(`Molekule: "${variant}" variant not found in theme...`);
  return config;
};

export const getComponentStyle = componentName => ({ theme }) => themeGet(theme, `components.${componentName}.style`);

export const getComponentClassName = ({ className, theme: { classPrefix }, variant }, name) =>
  `${className || ''} ${classPrefix}-${name} ${variant ? `${classPrefix}-${name}-${variant}` : ''}`.trim();

export const createComponent = ({ name, tag = 'div', as, style, props: defaultProps = () => ({}) }) => {
  const component = as ? styled(as) : styled[tag];

  return component.attrs(props => ({
    ...defaultProps(props),
    ...props,
    className: getComponentClassName(props, kebabCase(name)),
  }))`
    ${style}
    ${getComponentStyle(name)}
    ${({ styles = {} }) => styles[name] || {}}
  `;
};
