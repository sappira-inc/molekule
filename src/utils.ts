import kebabCase from 'lodash/kebabCase';
import get from 'lodash/get';
import styled, { CSSProp } from 'styled-components';

export const themeGet = (lookup: any, fallback?: any) => ({ theme }: any = {}) => get(theme, lookup, fallback);

export const getComponentVariant = (theme: any, componentName: string, variant: string) => {
  const config = themeGet(`variants.${componentName}.${variant}`)({
    theme,
  });

  if (!config) throw new Error(`Molekule: "${variant}" variant not found in theme...`);
  return config;
};

export const getComponentSize = (theme: any, componentName: string, size: string) =>
  themeGet(`sizes.${componentName}.${size}`, {})({ theme });

export const getComponentStyle = (componentName: string) => themeGet(`styles.${componentName}`, {});

export const getVariantStyles = (componentName: string, variant: string) =>
  themeGet(`variants.${componentName}.${variant}.style`, {});

interface GetComponentClassNameProps {
  className: string;
  theme: {
    classPrefix: string;
  };
  variant: string;
}

const getComponentClassName = (
  { className, theme: { classPrefix }, variant }: GetComponentClassNameProps,
  name: string
) => `${className || ''} ${classPrefix}-${name} ${variant ? `${classPrefix}-${name}-${variant}` : ''}`.trim();

interface CreateComponentProps<T> {
  name: string;
  tag?: keyof JSX.IntrinsicElements | React.JSXElementConstructor<any>;
  as?: React.ComponentType<any>;
  style?: ((props: any) => CSSProp) | CSSProp;
  props?: (props: T) => any;
}

export const createComponent = <
  T extends object,
  O extends keyof JSX.IntrinsicElements | React.ComponentType<any> = 'div'
>({
  name,
  tag = 'div',
  as,
  style,
  props: getBaseProps = () => ({}),
}: CreateComponentProps<T>) => {
  const component = styled<O>((as || tag) as any);

  return component.attrs<T>((props: any) => {
    const baseProps = getBaseProps(props);
    const finalProps = {
      ...baseProps,
      ...props,
      className: baseProps.className || props.className,
    };

    return {
      ...finalProps,
      className: getComponentClassName(finalProps, kebabCase(name)),
    };
  })<T>`
  ${style}
  ${getComponentStyle(name)}
  ${(p: any) => getVariantStyles(name, p.variant)}
  ${({ styles = {} }) => styles[name] || {}}
`;
};

// eslint-disable-next-line
const focusableFilter = (node: any) =>
  node.tabIndex >= 0 && !node.hasAttribute('disabled') ? NodeFilter.FILTER_ACCEPT : NodeFilter.FILTER_SKIP;

type FilterFn = (node?: any) => any;

const createTreeWalker = (root: any, currentNode: any, filterFn: FilterFn = () => NodeFilter.FILTER_ACCEPT) => {
  const treeWalker = document.createTreeWalker(
    root,
    NodeFilter.SHOW_ELEMENT,
    {
      acceptNode: filterFn,
    },
    false
  );
  treeWalker.currentNode = currentNode;
  return treeWalker;
};

export const findPreviousFocusableElement = (root: any, currentNode: any) => {
  const treeWalker = createTreeWalker(root, currentNode, focusableFilter);

  if (!treeWalker.previousNode() || treeWalker.currentNode === root) {
    return treeWalker.lastChild();
  }

  return treeWalker.currentNode;
};

export const findNextFocusableElement = (root: any, currentNode: any) => {
  const treeWalker = createTreeWalker(root, currentNode, focusableFilter);

  if (!treeWalker.nextNode() || treeWalker.currentNode === root) {
    treeWalker.currentNode = root;
    return treeWalker.firstChild();
  }

  return treeWalker.currentNode;
};

export const getNextCursorPosition = (cursorPos: any, oldValue: any, newValue: any) => {
  const cursorDiff = newValue.length - oldValue.length;
  let nextPosition = cursorPos;

  if (cursorDiff > 1 || cursorDiff < -1) {
    nextPosition += cursorDiff < 0 ? cursorDiff + 1 : cursorDiff - 1;
  }

  return Math.max(nextPosition, 0);
};
