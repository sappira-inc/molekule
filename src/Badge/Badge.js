import React from 'react';
import PropTypes from 'prop-types';
import { css } from 'styled-components';
import { space } from 'styled-system';
import propTypes from '@styled-system/prop-types';
import { getComponentVariant, createComponent, getComponentSize } from '../utils';

const StyledBadge = createComponent({
  name: 'Badge',
  tag: 'span',
  style: ({ variant, theme, size }) => {
    const variantStyles = getComponentVariant(theme, 'Badge', variant);
    const sizeStyles = getComponentSize(theme, 'Badge', size);

    return css`
      font-family: inherit;
      font-weight: bold;

      ${variantStyles};
      ${sizeStyles};
      ${space};
    `;
  },
});

const Badge = React.forwardRef((props, ref) => <StyledBadge {...props} ref={ref} />);

Badge.propTypes = {
  variant: PropTypes.string,
  size: PropTypes.string,
  ...propTypes.space,
};

Badge.defaultProps = {
  variant: 'info',
  size: 'md',
};

export default Badge;
