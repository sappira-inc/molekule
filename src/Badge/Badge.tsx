import React from 'react';
import { css } from 'styled-components';
import { space, SpaceProps } from 'styled-system';
import { getComponentVariant, createComponent, getComponentSize } from '../utils';

export interface BadgeProps extends SpaceProps {
  variant?: string;
  size?: string;
  children?: React.ReactNode;
}

export const Badge = createComponent<BadgeProps, 'span'>({
  name: 'Badge',
  tag: 'span',
  style: ({ variant, theme, size }) => {
    const variantStyles = variant ? getComponentVariant(theme, 'Badge', variant) : '';
    const sizeStyles = getComponentSize(theme, 'Badge', size);

    return css`
      font-family: inherit;
      font-weight: bold;

      ${variantStyles}
      ${sizeStyles}
      ${space}
    `;
  },
});

Badge.defaultProps = {
  variant: 'info',
  size: 'md',
};
