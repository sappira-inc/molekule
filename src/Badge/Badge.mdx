---
name: Badge
menu: Core
---

import { Playground, PropsTable } from 'docz';
import Badge from './Badge';

# Badge

Documentation and examples for badges, our small count and labeling component.

## Properties

<PropsTable of={Badge} />

## Examples

<Playground>
  <Badge>I'm Ze Best</Badge>
  <Badge variant="success" ml={2}>I'm So Successful</Badge>
  <Badge variant="danger" ml={2}>I'm So Dangerous</Badge>
  <Badge variant="warning" ml={2}>💁‍♀️ Caution</Badge>
</Playground>
