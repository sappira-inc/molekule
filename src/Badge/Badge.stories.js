import React from 'react';
import Badge from './Badge';

export default {
  title: 'Components|Badge',
  component: Badge,
};

export const Basic = () => <Badge>Default Badge</Badge>;

export const Variants = () => (
  <>
    <Badge>Default</Badge>
    <Badge variant="success" ml={2}>
      Success
    </Badge>
    <Badge variant="danger" ml={2}>
      Danger
    </Badge>
    <Badge variant="warning" ml={2}>
      Caution
    </Badge>
  </>
);

export const Sizes = () => (
  <>
    <Badge size="sm">Small</Badge>
    <Badge ml={2}>Medium</Badge>
    <Badge size="lg" ml={2}>
      Large
    </Badge>
  </>
);
