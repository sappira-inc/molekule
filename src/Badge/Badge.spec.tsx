import React from 'react';
import { renderWithTheme } from '../test/utils';
import Badge from './Badge';

test('Badge', () => {
  const { asFragment } = renderWithTheme(<Badge>I&apos;m a badge!</Badge>);

  expect(asFragment()).toMatchSnapshot();
});
