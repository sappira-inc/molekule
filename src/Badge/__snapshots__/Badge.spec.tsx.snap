// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge 1`] = `
<DocumentFragment>
  .c0 {
  font-family: inherit;
  font-weight: bold;
}

<span
    class="c0 re-badge"
  >
    I'm a badge!
  </span>
</DocumentFragment>
`;
